File: SRC/zhetri2.f
      SUBROUTINE ZHETRI2( UPLO, N, A, LDA, IPIV, WORK, LWORK, INFO )

*  -- LAPACK computational routine --
*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--

      // .. Scalar Arguments ..
      String             UPLO;
      int                INFO, LDA, LWORK, N;
      // ..
      // .. Array Arguments ..
      int                IPIV( * );
      COMPLEX*16         A( LDA, * ), WORK( * )
      // ..

*  =====================================================================

      // .. Local Scalars ..
      bool               UPPER, LQUERY;
      int                MINSIZE, NBMAX;
      // ..
      // .. External Functions ..
      bool               LSAME;
      int                ILAENV;
      // EXTERNAL LSAME, ILAENV
      // ..
      // .. External Subroutines ..
      // EXTERNAL ZHETRI2X, ZHETRI, XERBLA
      // ..
      // .. Executable Statements ..

      // Test the input parameters.

      INFO = 0
      UPPER = LSAME( UPLO, 'U' )
      LQUERY = ( LWORK == -1 )

      // Get blocksize

      NBMAX = ILAENV( 1, 'ZHETRF', UPLO, N, -1, -1, -1 )
      if ( N == 0 ) {
         MINSIZE = 1
      } else if ( NBMAX.GE.N ) {
         MINSIZE = N
      } else {
         MINSIZE = (N+NBMAX+1)*(NBMAX+3)
      }

      if ( .NOT.UPPER .AND. .NOT.LSAME( UPLO, 'L' ) ) {
         INFO = -1
      } else if ( N.LT.0 ) {
         INFO = -2
      } else if ( LDA.LT.MAX( 1, N ) ) {
         INFO = -4
      } else if ( LWORK.LT.MINSIZE .AND. .NOT.LQUERY ) {
         INFO = -7
      }

      if ( INFO.NE.0 ) {
         xerbla('ZHETRI2', -INFO );
         RETURN
      } else if ( LQUERY ) {
         WORK( 1 ) = MINSIZE
         RETURN
      }

      // Quick return if possible

      if (N == 0) RETURN;

      if ( NBMAX.GE.N ) {
         zhetri(UPLO, N, A, LDA, IPIV, WORK, INFO );
      } else {
         zhetri2x(UPLO, N, A, LDA, IPIV, WORK, NBMAX, INFO );
      }

      RETURN

      // End of ZHETRI2

      }

File: lib/src/zhetri2.dart
// Copyright (c) 2024 Guilherme Lepsch. All rights reserved. Use of this
// source code is governed by a BSD-style license that can be found in the
// [LICENSE file](https://github.com/lepsch/dart_lapack/blob/main/LICENSE).

import 'dart:math';

import 'package:dart_lapack/src/install/lsame.dart';
import 'package:dart_lapack/src/box.dart';
import 'package:dart_lapack/src/complex.dart';
import 'package:dart_lapack/src/ilaenv.dart';
import 'package:dart_lapack/src/matrix.dart';
import 'package:dart_lapack/src/xerbla.dart';
import 'package:dart_lapack/src/zhetri.dart';
import 'package:dart_lapack/src/zhetri2x.dart';

void zhetri2(
  final String UPLO,
  final int N,
  final Matrix<Complex> A_,
  final int LDA,
  final Array<int> IPIV_,
  final Array<Complex> WORK_,
  final int LWORK,
  final Box<int> INFO,
) {
  final A = A_.having(ld: LDA);
  final IPIV = IPIV_.having();
  final WORK = WORK_.having();
  bool UPPER, LQUERY;
  int MINSIZE, NBMAX;

  // Test the input parameters.

  INFO.value = 0;
  UPPER = lsame(UPLO, 'U');
  LQUERY = (LWORK == -1);

  // Get blocksize

  NBMAX = ilaenv(1, 'ZHETRF', UPLO, N, -1, -1, -1);
  if (N == 0) {
    MINSIZE = 1;
  } else if (NBMAX >= N) {
    MINSIZE = N;
  } else {
    MINSIZE = (N + NBMAX + 1) * (NBMAX + 3);
  }

  if (!UPPER && !lsame(UPLO, 'L')) {
    INFO.value = -1;
  } else if (N < 0) {
    INFO.value = -2;
  } else if (LDA < max(1, N)) {
    INFO.value = -4;
  } else if (LWORK < MINSIZE && !LQUERY) {
    INFO.value = -7;
  }

  if (INFO.value != 0) {
    xerbla('ZHETRI2', -INFO.value);
    return;
  } else if (LQUERY) {
    WORK[1] = MINSIZE.toComplex();
    return;
  }

  // Quick return if possible

  if (N == 0) return;

  if (NBMAX >= N) {
    zhetri(UPLO, N, A, LDA, IPIV, WORK, INFO);
  } else {
    zhetri2x(UPLO, N, A, LDA, IPIV, WORK.asMatrix(), NBMAX, INFO);
  }
}
